### PR TITLE
#285: add support for multi-line headers (RqLive)

### PR DESCRIPTION
**Pull request by lautarobock**
#285: Checks if a new line starts with SP and has not colon. Then appends it to last header value.

### CODE REVIEW

**Comment by davvd** [comment]
@lautarobock Many thanks, I will find someone to review it before we merge

**Comment by davvd** [comment]
@carlosmiranda please review, thanks

File: src/main/java/org/takes/rq/RqLive.java [hunk 1]
```diff
@@ -63,7 +63,7 @@ public RqLive(final InputStream input) throws IOException {
      */
     @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
     private static Request parse(final InputStream input) throws IOException {
-        final Collection<String> head = new LinkedList<String>();
+        final List<String> head = new LinkedList<String>();
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rq/RqLive.java]: @lautarobock Why was this changed to `List`? I don't see any `List` specific operations.

File: src/main/java/org/takes/rq/RqLive.java [hunk 2]
```diff
@@ -84,12 +84,13 @@ private static Request parse(final InputStream input) throws IOException {
                 if (baos.size() == 0) {
                     break;
                 }
-                head.add(new String(baos.toByteArray()));
+                final String header = new String(baos.toByteArray());
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rq/RqLive.java]: @lautarobock No need for this `header` variable, let's inline it with the method call below.

File: src/main/java/org/takes/rq/RqLive.java [hunk 3]
```diff
@@ -113,4 +114,29 @@ public InputStream body() {
...
+     * @param header Current header
+     * @throws HttpException if multiline do not start with space
+     */
+    private static void checkAndAddHeader(final List<String> head,
+            final String header) throws HttpException {
+        if (head.isEmpty() || header.contains(":")) {
+            head.add(header);
+        } else if (header.charAt(0) == ' ' || header.charAt(0) == '\t') {
+            final String prev = head.remove(head.size() - 1);
+            head.add(prev + header);
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rq/RqLive.java]: @lautarobock Let's [avoid string concatenation[(http://www.yegor256.com/2014/06/19/avoid-string-concatenation.html). Also, we can join the previous line with this, avoiding the single use variable.

File: src/main/java/org/takes/rq/RqLive.java [hunk 4]
```diff
@@ -113,4 +114,29 @@ public InputStream body() {
...
     }
 
+    /**
+     * Check line and add header o replace last one.
+     * @param head All previous headers
+     * @param header Current header
+     * @throws HttpException if multiline do not start with space
+     */
+    private static void checkAndAddHeader(final List<String> head,
+            final String header) throws HttpException {
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rq/RqLive.java]: @lautarobock Too much indentation here, should be back four spaces.

**Comment by carlosmiranda** [comment]
@lautarobock Some comments above. Also, please add a description to the pull request [as per our QA policy](http://www.teamed.io/qa.html).

File: src/main/java/org/takes/rq/RqLive.java (hunk 1 above)
Comment (lautarobock) [inline src/main/java/org/takes/rq/RqLive.java]: in checkAndAddHeader method use List.remove() method

**Comment by lautarobock** [comment]
@carlosmiranda PR is updated. New commit and description

File: src/main/java/org/takes/rq/RqLive.java [hunk 5]
```diff
@@ -113,4 +113,32 @@ public InputStream body() {
...
 
+    /**
+     * Check line and add header o replace last one.
+     * @param head All previous headers
+     * @param header Current header
+     * @throws HttpException if multiline do not start with space
+     */
+    private static void checkAndAddHeader(
+        final List<String> head,
+        final String header
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rq/RqLive.java]: @lautarobock Minor comment - let's move this param up to line 123

File: src/main/java/org/takes/rq/RqLive.java (hunk 1 above)
Comment (carlosmiranda) [inline src/main/java/org/takes/rq/RqLive.java]: @lautarobock Makes sense, thanks. Btw, please address me in the previous comment with @carlosmiranda. All messages should start with someone's name.

**Comment by carlosmiranda** [comment]
@lautarobock Just one more comment. Also, please add a link to the original issue in your PR description.

**Comment by lautarobock** [comment]
@carlosmiranda PR is updated

**Comment by carlosmiranda** [comment]
@lautarobock Thanks!

**Comment by carlosmiranda** [comment]
@rultor merge please

**Comment by yegor256** [comment]
@lautarobock this implementation is too complex (can be done much simpler) and wrong. This request won't work correctly:

```
GET / HTTP/1.1
Foo: first line
 second: line
```

**Comment by yegor256** [comment]
@lautarobock the idea with `List` is wrong, it's a rather procedural style. Instead, I think we should stay in the same method and just manipulate with `baos`

**Comment by lautarobock** [comment]
@yegor256 List was removed. Fix the bug. Reduce complexity

**Comment by yegor256** [comment]
@lautarobock omg, now it's even more complex :) why can't we do it just a few extra lines inside existing ctor of `RqLive`? looks like you're over-engineering here

**Comment by lautarobock** [comment]
@yegor256 yes, but I'need to "peek" the next char after end of line. And I don't want to use not final "data" or a boolean for checking if it need to read one more char or not.

**Comment by lautarobock** [comment]
@yegor256  Besides. the if with an AND (&&) increase the complexity over 10. Thats way I have to extract comparation in a class.
If you dont't mind to use extra method for comparison and use not fina "data", I can do a simpler solution

**Comment by lautarobock** [comment]
@yegor256 There is a new and simpler solution

**Comment by lautarobock** [comment]
@yegor256 why the buld still yellow?

**Comment by yegor256** [comment]
@lautarobock looks better now, but I still don't understand what class `StartingMLCond` is for? you're using it only once. why can't you just use `if` where it's needed?

**Comment by lautarobock** [comment]
@yegor256 is for keep complexity under 10. the "if" with two conditions increases its complexity

**Comment by yegor256** [comment]
@lautarobock let's suppress Checkstyle and PMD, if they complain about complexity. This solution (with a supplementary class) makes code less readable. To properly solve the complexity problem we should refactor the entire method somehow

**Comment by lautarobock** [comment]
@yegor256  Done

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@ypshenychka please, review this task for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor please deploy

**Comment by ypshenychka** [comment]
@davvd Please assign @carlosmiranda to this PR as far as he was code reviewer.

**Comment by carlosmiranda** [comment]
@ypshenychka I am assigned to this PR as indicated by `@carlosmiranda` label.

**Comment by ypshenychka** [comment]
@carlosmiranda sure, I see. Just wanted to make sure that you will be paid for this ticket.

**Comment by ypshenychka** [comment]
@davvd Quality is good.

**Comment by carlosmiranda** [comment]
@ypshenychka Thanks for the concern. :D

**Comment by davvd** [comment]
> @davvd Quality is good.

@ypshenychka thanks, got it, great to see high-quality tickets here

**Comment by davvd** [comment]
@carlosmiranda since quality is good, I just added 10 mins to @yegor256 (our architect) in transaction 000-2b041c6a

**Comment by davvd** [comment]
@carlosmiranda I added 10 mins to @ypshenychka (for QA review) in transaction 57789341... **21 mins** sent to your balance (ID `57789348`), many thanks! It took 95 hours and 25 mins.... you're getting extra minutes here (c=6)... +21 to your rating, your total score is [+4208](http://www.netbout.com/b/35023?open=rating)
